File: sdks/ruby/lib/patala/ffi.rb
# frozen_string_literal: true

# patala in-process, through the C ABI (`libpatala_ffi`), using `fiddle` from
# the standard library. No child process, no port, no gem dependency.
#
#   require "patala/ffi"
#
#   Patala::Ffi.open do |rail|                     # MockRail by default
#     receipt = rail.charge(amount_minor: 1250, currency: "USDC",
#                           destination: "mock:wallet:alice", reference: "order-1")
#     rail.verify(receipt).fetch("valid")          # => true
#   end
#
# Why `fiddle` and not the `ffi` gem: fiddle ships with Ruby, so direct mode
# adds no dependency at all. Everything needed is there — dlopen and a
# struct-free calling convention. patala has no streaming operation, so there
# is no callback to arrange either; six functions is the whole ABI.
#
# WHAT LOADING THIS COSTS YOU: almost nothing, and that is not the usual answer.
# patala is Rust, so no language runtime enters your process — no GC, no
# scheduler, no signal handlers, and no threads (each handle owns a
# current-thread tokio runtime that runs on whichever thread called in). If you
# have read the equivalent file in llmux or openrate, its Go-runtime warnings
# are true there and false here; see ../../README.md, where it is measured.
#
# The ONE fork rule, which is real: a handle's runtime sits behind a mutex, and
# fork() copies a locked mutex as locked. Open handles in the child. Unicorn's
# `after_fork`, Puma's `on_worker_boot`, Resque's per-job child — build the
# Patala::Ffi there, not at boot. Loading the LIBRARY before the fork is fine;
# it is the handle that matters. examples/fork_probe.rb measures both.

require "fiddle"
require "json"
require "rbconfig"

module Patala
  class Error < StandardError; end unless defined?(Error)

  # The C ABI of libpatala_ffi (patala-ffi/include/patala.h), bound with fiddle.
  class Ffi
    PTR  = Fiddle::TYPE_VOIDP
    U64  = Fiddle::TYPE_LONG_LONG # fiddle has no unsigned 64-bit constant;
    INT  = Fiddle::TYPE_INT       # handles are small counters, so this is exact.
    VOID = Fiddle::TYPE_VOID

    # Every method `patala_call` dispatches, as of this ABI. Listed so a typo
    # is caught here with the real list attached rather than after a round trip
    # into the library.
    METHODS = %w[
      id capabilities quote charge verify validate-destination webhook caveat providers
    ].freeze

    # Open a rail, yield it, and close it however the block exits. Prefer this
    # to a bare .new: `ensure` around one is the same thing written longhand and
    # easier to get wrong.
    #
    #   Patala::Ffi.open { |rail| ... }                                  # MockRail
    #   Patala::Ffi.open(config: { rail: "mock", failing: true }) { ... } # every op fails
    def self.open(config: nil, library: nil)
      rail = new(config: config, library: library)
      begin
        yield rail
      ensure
        rail.close
      end
    end

    # config:  a rail configuration document — a Hash, a JSON String, or nil for
    #          the offline default (a deterministic MockRail on USDC). Unknown
    #          fields are REFUSED, not ignored: a misspelled "currencys" is an
    #          error rather than a rail quietly built with defaults you did not
    #          choose.
    # library: override library resolution entirely.
    def initialize(config: nil, library: nil)
      @library_path = library || self.class.resolve_library
      begin
        @dl = Fiddle.dlopen(@library_path)
      rescue Fiddle::DLError => e
        raise Error, "could not load #{@library_path}: #{e.message}"
      end

      bind!

      json = case config
             when nil, String then config
             else JSON.generate(config)
             end

      slot = err_slot
      handle = @fn_new.call(json, slot)
      # patala_new returns 0 for FAILURE, inverting the usual convention,
      # because its success value is a handle and handles start at 1.
      raise Error, "patala_new: #{take_error(slot)}" if handle.zero?

      @handle = handle
      @mutex = Mutex.new
    end

    # The path this instance actually loaded.
    attr_reader :library_path

    # The patala version the LOADED library was built from.
    #
    # Worth checking at startup: a shared library is resolved off a load path
    # you may not control, and a stale libpatala_ffi earlier on it is called
    # silently otherwise. The string is static inside the library — never freed.
    def version
      @fn_abi_version.call.to_s
    end

    # Raise unless the loaded library is exactly `expected`. The comparison
    # lives in the library (`patala_abi_check`) so it is not reimplemented —
    # and forgotten — in each binding.
    def abi_check!(expected)
      slot = err_slot
      return true if @fn_abi_check.call(expected.to_s, slot).zero?

      raise Error, take_error(slot)
    end

    # One unary call. Returns the parsed JSON response.
    #
    #   rail.call("charge", amount_minor: 1250, currency: "USDC", ...)
    #   rail.call("capabilities")
    def call(method, request = nil, **kwargs)
      request = kwargs unless kwargs.empty?
      JSON.parse(call_raw(method, request))
    end

    # The same call, returning the response JSON verbatim — the same bytes
    # POST /v1/rails/:id/<method> returns from patala-sidecar.
    def call_raw(method, request = nil)
      assert_open!
      method = method.to_s
      unless METHODS.include?(method)
        raise Error, "unknown method #{method.inspect} (want one of: #{METHODS.join(', ')})"
      end

      json = case request
             when nil, String then request
             else JSON.generate(request)
             end

      # A FRESH, zeroed slot per call. patala writes *err on failure only and
      # leaves it untouched otherwise, so a reused slot still holds the previous
      # (already freed) pointer.
      slot = err_slot
      res = @fn_call.call(@handle, method, json, slot)
      raise Error, "patala_call(#{method}): #{take_error(slot)}" if res.null?

      begin
        res.to_s # Fiddle::Pointer#to_s copies up to the NUL into a Ruby String
      ensure
        # Everything the library returns — results AND error messages — is
        # released with patala_free and with nothing else. Not Ruby's free,
        # not yours: this is Rust's allocator.
        @fn_free.call(res)
      end
    end

    # ------------------------------------------------------------- convenience

    def id           = call("id").fetch("rail_id")
    def capabilities = call("capabilities")
    def quote(request = nil, **kw) = call("quote", request, **kw)
    def charge(request = nil, **kw) = call("charge", request, **kw)

    # Hand back the Receipt "charge" returned, unchanged. Returns the whole
    # {"valid" => bool} document rather than the bool, because `false` here is
    # the rail's honest fail-closed answer and not an error — gate entitlement
    # on `true` and never retry a `false` as though it were transient.
    def verify(receipt) = call("verify", receipt)

    # The offline pre-flight check. NEVER raises for an unknown address: "I
    # cannot check this" comes back as the verdict {"status" => "Unknown"},
    # because a caller must handle it as carefully as a refusal and an error is
    # too easy to swallow. Read "is_refusal" (do not send) and
    # "human_must_confirm" (true on EVERY verdict, "StructurallyValid" included).
    def validate_destination(destination)
      call("validate-destination", destination: destination)
    end

    # Authenticate an inbound processor delivery. Forward it VERBATIM — the
    # exact body bytes, the headers, the query string — because every scheme
    # signs what the processor actually sent, and a body that has been through
    # a JSON round trip on your side is no longer that.
    def webhook(body:, headers: {}, query: {}, now_unix: Time.now.to_i, binary: false)
      payload = { headers: headers, query: query, now_unix: now_unix }
      payload[binary ? :body_hex : :body] = binary ? body.unpack1("H*") : body
      call("webhook", payload)
    end

    # The sentence to show a human on the form where a payout address is first
    # asked for — before there is a verdict to render. patala does not detect
    # exchange-owned addresses and this is how it says so.
    def caveat = call("caveat").fetch("exchange_deposit_caveat")

    # Fiat provider names this specific build can construct. Only present when
    # the library was built with --features fiat.
    def providers = call("providers").fetch("providers")

    # Release the rail. Idempotent, so cleanup paths can call it blindly.
    def close
      @mutex&.synchronize do
        return if @handle.nil? || @handle.zero?

        @fn_close.call(@handle)
        @handle = 0
      end
      nil
    end

    def closed? = @handle.nil? || @handle.zero?

    # ---------------------------------------------------------------- internals

    private

    def bind!
      @fn_abi_version = Fiddle::Function.new(@dl["patala_abi_version"], [], PTR,
                                             name: "patala_abi_version")
      @fn_abi_check = Fiddle::Function.new(@dl["patala_abi_check"], [PTR, PTR], INT,
                                           name: "patala_abi_check")
      @fn_new = Fiddle::Function.new(@dl["patala_new"], [PTR, PTR], U64, name: "patala_new")
      @fn_close = Fiddle::Function.new(@dl["patala_close"], [U64], VOID, name: "patala_close")
      @fn_call = Fiddle::Function.new(@dl["patala_call"], [U64, PTR, PTR, PTR], PTR,
                                      name: "patala_call")
      @fn_free = Fiddle::Function.new(@dl["patala_free"], [PTR], VOID, name: "patala_free")
    rescue Fiddle::DLError => e
      raise Error, "#{@library_path} is missing a symbol the ABI promises: #{e.message}"
    end

    def assert_open!
      raise Error, "this Patala::Ffi handle is closed" if closed?
    end

    # A zeroed `char*` slot for the trailing `char** err` argument.
    def err_slot
      slot = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP, Fiddle::RUBY_FREE)
      slot[0, Fiddle::SIZEOF_VOIDP] = "\x00".b * Fiddle::SIZEOF_VOIDP
      slot
    end

    # Read the message out of an err slot and free it with patala_free — an
    # error must not also be a leak.
    def take_error(slot)
      ptr = slot.ptr
      return "(no message)" if ptr.null?

      message = ptr.to_s
      @fn_free.call(ptr)
      message
    end

    class << self
      # 1. PATALA_LIBRARY
      # 2. lib/ inside this gem (where a platform gem would put it)
      # 3. target/{debug,release}/ in a checkout of the patala workspace
      # 4. the bare soname, letting the dynamic loader search its own paths
      def resolve_library
        env = ENV["PATALA_LIBRARY"]
        return env if env && !env.empty?

        gem_root = File.expand_path("../..", __dir__) # …/sdks/ruby
        repo = File.expand_path("../..", gem_root)    # …/repo root

        candidates = [File.join(gem_root, "lib", library_name)]
        %w[debug release].each do |profile|
          candidates << File.join(repo, "target", profile, library_name)
        end
        candidates.find { |path| File.file?(path) } || library_name
      end

      def library_name
        case RbConfig::CONFIG["host_os"]
        when /mswin|mingw|cygwin/ then "patala_ffi.dll"
        when /darwin/ then "libpatala_ffi.dylib"
        else "libpatala_ffi.so"
        end
      end
    end
  end
end

File: sdks/php/src/Ffi.php
<?php

declare(strict_types=1);

namespace Patala;

/**
 * patala in-process, through the C ABI (`libpatala_ffi`), using PHP's bundled
 * FFI extension. No child process, no port.
 *
 *   $rail = new \Patala\Ffi();                    // offline MockRail
 *   try {
 *       $receipt = $rail->charge([
 *           'amount_minor' => 1250, 'currency' => 'USDC',
 *           'destination'  => 'mock:wallet:alice', 'reference' => 'order-1',
 *       ]);
 *       $ok = $rail->verify($receipt)['valid'];   // true
 *   } finally {
 *       $rail->close();                            // frees the handle, always
 *   }
 *
 * WHAT LOADING THIS COSTS YOU: almost nothing, and that is not the usual
 * answer. patala is Rust — no language runtime enters your process, no GC, no
 * scheduler, no signal handlers, and no threads. If you have read the
 * equivalent file in llmux or openrate, its warning about php-fpm and the
 * non-fork-safe Go runtime is true there and FALSE here; see ../README.md,
 * where it is measured against a real forking parent rather than asserted.
 *
 * The one rule that IS real: open a handle in the child. A handle's runtime
 * sits behind a mutex and fork() copies a locked mutex as locked, so a handle
 * built in the php-fpm master and used in a worker can deadlock. Build it
 * per-request, or per-worker after the fork — which is what almost all PHP
 * code does anyway.
 *
 * Requirements: ext-ffi (bundled with PHP >= 7.4, but see `ffi.enable`) and a
 * libpatala_ffi built for this platform (`cargo build -p patala-ffi`).
 */
final class Ffi
{
    /**
     * The six functions of the C ABI, transcribed from
     * patala-ffi/include/patala.h. PHP's FFI parser does not run the C
     * preprocessor, so the header itself (with its #ifndef and #include)
     * cannot be handed to FFI::cdef.
     */
    private const CDEF = <<<'C'
        const char* patala_abi_version(void);
        int patala_abi_check(const char* expected, char** err);
        uint64_t patala_new(const char* config_json, char** err);
        void patala_close(uint64_t h);
        char* patala_call(uint64_t h, const char* method, const char* request_json, char** err);
        void patala_free(char* p);
        C;

    /**
     * Every method `patala_call` dispatches. Checked here so a typo is caught
     * with the real list attached rather than after a round trip.
     */
    public const METHODS = [
        'id', 'capabilities', 'quote', 'charge', 'verify',
        'validate-destination', 'webhook', 'caveat', 'providers',
    ];

    /** @var \FFI */
    private $ffi;

    /** @var int */
    private $handle = 0;

    /** @var string */
    private $libraryPath;

    /**
     * @param array<string,mixed>|string|null $config a rail configuration
     *        document. null means the offline default: a deterministic
     *        MockRail on USDC, no credentials and no network. Unknown fields
     *        are REFUSED rather than ignored.
     * @param string|null $libraryPath override library resolution entirely.
     */
    public function __construct($config = null, ?string $libraryPath = null)
    {
        if (!\extension_loaded('FFI')) {
            throw new PatalaException(
                'the FFI extension is not loaded. It ships with PHP >= 7.4 but is ' .
                'often disabled: enable extension=ffi in php.ini. See sdks/php/README.md.'
            );
        }

        $this->libraryPath = $libraryPath ?? self::resolveLibrary();

        try {
            $this->ffi = \FFI::cdef(self::CDEF, $this->libraryPath);
        } catch (\FFI\Exception $e) {
            throw new PatalaException(
                'FFI::cdef failed for ' . $this->libraryPath . ': ' . $e->getMessage() .
                ($this->looksRestricted($e)
                    ? "\nffi.enable is 'preload' by default, which forbids FFI::cdef in " .
                      'any SAPI except CLI. Either preload a file that opens the library, ' .
                      'set ffi.enable=true, or use \\Patala\\Sidecar — which needs no ' .
                      'php.ini change at all. See sdks/php/README.md.'
                    : ''),
                0,
                $e
            );
        }

        $json = \is_string($config) || $config === null ? $config : self::encode($config);

        $err = $this->ffi->new('char*');
        // patala_new returns 0 for FAILURE — its success value is a handle,
        // and handles start at 1.
        $handle = $this->ffi->patala_new($json, \FFI::addr($err));
        if ($handle === 0) {
            throw new PatalaException('patala_new: ' . $this->takeError($err));
        }
        $this->handle = (int) $handle;
    }

    /**
     * Run $fn with an open rail and close it afterwards, whatever happens.
     * PHP has no `using`/`with`, so this is the closest idiomatic construct;
     * try/finally around a plain constructor is equally fine.
     *
     * @param array<string,mixed>|string|null $config
     * @return mixed whatever $fn returns
     */
    public static function with($config, callable $fn, ?string $libraryPath = null)
    {
        $rail = new self($config, $libraryPath);
        try {
            return $fn($rail);
        } finally {
            $rail->close();
        }
    }

    /**
     * The patala version the LOADED library was built from.
     *
     * Worth checking at startup: a shared library is resolved off a load path
     * you may not control, and a stale libpatala_ffi earlier on it is called
     * silently otherwise. The string is static in the library — never freed.
     */
    public function version(): string
    {
        return (string) $this->ffi->patala_abi_version();
    }

    /**
     * Throw unless the loaded library is exactly $expected. The comparison
     * lives in the library so it is not reimplemented — and forgotten — in
     * each binding.
     */
    public function abiCheck(string $expected): void
    {
        $err = $this->ffi->new('char*');
        if ($this->ffi->patala_abi_check($expected, \FFI::addr($err)) !== 0) {
            throw new PatalaException($this->takeError($err));
        }
    }

    public function libraryPath(): string
    {
        return $this->libraryPath;
    }

    /**
     * One unary call, decoded.
     *
     * @param array<string,mixed>|string|null $request
     * @return array<string,mixed>
     */
    public function call(string $method, $request = null): array
    {
        $raw = $this->callRaw($method, $request);
        $decoded = \json_decode($raw, true);
        if (!\is_array($decoded)) {
            throw new PatalaException("patala returned a non-object response to '{$method}'");
        }

        return $decoded;
    }

    /**
     * The same call, returning the response JSON verbatim — the same bytes
     * `POST /v1/rails/:id/<method>` returns from patala-sidecar.
     *
     * @param array<string,mixed>|string|null $request
     */
    public function callRaw(string $method, $request = null): string
    {
        $this->assertOpen();
        if (!\in_array($method, self::METHODS, true)) {
            throw new PatalaException(
                "unknown method '{$method}' (want one of: " . \implode(', ', self::METHODS) . ')'
            );
        }

        $json = \is_string($request) || $request === null ? $request : self::encode($request);

        // A FRESH slot per call, deliberately. patala writes *err on failure
        // only and leaves it untouched otherwise, so a reused slot still holds
        // the previous (already-freed) pointer. FFI::new zeroes its allocation.
        $err = $this->ffi->new('char*');
        $res = $this->ffi->patala_call($this->handle, $method, $json, \FFI::addr($err));

        // A NULL `char*` return arrives in PHP as null, not as a null CData.
        if ($res === null) {
            throw new PatalaException("patala_call({$method}): " . $this->takeError($err));
        }

        try {
            return \FFI::string($res);
        } finally {
            // Everything the library returns — results AND error messages — is
            // released with patala_free and nothing else. Not PHP's memory
            // manager, not free(): this is Rust's allocator.
            $this->ffi->patala_free($res);
        }
    }

    // --------------------------------------------------------- convenience

    public function id(): string
    {
        return $this->call('id')['rail_id'];
    }

    /** @return array<string,mixed> */
    public function capabilities(): array
    {
        return $this->call('capabilities');
    }

    /**
     * @param array<string,mixed> $request
     * @return array<string,mixed>
     */
    public function quote(array $request): array
    {
        return $this->call('quote', $request);
    }

    /**
     * @param array<string,mixed> $request
     * @return array<string,mixed> the Receipt — store it and hand it back to
     *         verify() later. THAT, not charge() having returned, is the
     *         entitlement check.
     */
    public function charge(array $request): array
    {
        return $this->call('charge', $request);
    }

    /**
     * @param array<string,mixed> $receipt the Receipt charge() returned
     * @return array{valid:bool} the whole document, not the bool: `false` is
     *         the rail's fail-closed answer and is data, not an error.
     */
    public function verify(array $receipt): array
    {
        return $this->call('verify', $receipt);
    }

    /**
     * The offline pre-flight check. NEVER throws for an unknown address: "I
     * cannot check this" comes back as {"status": "Unknown"}. Read
     * `is_refusal` (do not send) and `human_must_confirm` (true on EVERY
     * verdict, "StructurallyValid" included).
     *
     * @return array<string,mixed>
     */
    public function validateDestination(string $destination): array
    {
        return $this->call('validate-destination', ['destination' => $destination]);
    }

    /**
     * Authenticate an inbound processor delivery. Forward it VERBATIM — the
     * exact body bytes, the headers, the query string — because every scheme
     * signs what the processor actually sent. A body that has been through
     * json_decode/json_encode on your side is no longer that.
     *
     * @param array<string,string> $headers
     * @param array<string,string> $query
     * @return array<string,mixed> a WebhookEvent
     */
    public function webhook(
        string $body,
        array $headers = [],
        array $query = [],
        ?int $nowUnix = null,
        bool $binary = false
    ): array {
        $payload = [
            'headers' => (object) $headers,
            'query' => (object) $query,
            'now_unix' => $nowUnix ?? \time(),
        ];
        $payload[$binary ? 'body_hex' : 'body'] = $binary ? \bin2hex($body) : $body;

        return $this->call('webhook', $payload);
    }

    /**
     * The sentence to show a human on the form where a payout address is first
     * asked for. patala does not detect exchange-owned addresses, and this is
     * how it says so.
     */
    public function caveat(): string
    {
        return $this->call('caveat')['exchange_deposit_caveat'];
    }

    /**
     * Fiat provider names this build can construct. Only present when the
     * library was built with --features fiat.
     *
     * @return list<string>
     */
    public function providers(): array
    {
        return $this->call('providers')['providers'];
    }

    /** Release the rail. Idempotent, so cleanup paths can call it blindly. */
    public function close(): void
    {
        if ($this->handle !== 0) {
            $this->ffi->patala_close($this->handle);
            $this->handle = 0;
        }
    }

    /**
     * A safety net, not the mechanism. PHP destroys objects at unpredictable
     * points and not at all on a fatal error; close() in a finally block is
     * what actually guarantees the handle goes away.
     */
    public function __destruct()
    {
        $this->close();
    }

    // ------------------------------------------------------------ internals

    private function assertOpen(): void
    {
        if ($this->handle === 0) {
            throw new PatalaException('this Patala\Ffi handle is closed');
        }
    }

    /** @param \FFI\CData $err a `char*` slot */
    private function takeError($err): string
    {
        if (\FFI::isNull($err)) {
            return '(no message)';
        }
        $message = \FFI::string($err);
        $this->ffi->patala_free($err);

        return $message;
    }

    private function looksRestricted(\FFI\Exception $e): bool
    {
        return \strpos($e->getMessage(), 'ffi.enable') !== false
            || \strpos($e->getMessage(), 'FFI API is restricted') !== false;
    }

    /** @param array<string,mixed> $value */
    private static function encode(array $value): string
    {
        $json = \json_encode($value);
        if ($json === false) {
            throw new PatalaException('could not encode the request as JSON: ' . \json_last_error_msg());
        }

        return $json;
    }

    /**
     * Find libpatala_ffi:
     *   1. PATALA_LIBRARY
     *   2. lib/ inside this package
     *   3. target/{debug,release}/ in a checkout of the patala workspace
     *   4. the bare soname, letting the dynamic loader search its own paths
     */
    private static function resolveLibrary(): string
    {
        $env = \getenv('PATALA_LIBRARY');
        if ($env !== false && $env !== '') {
            return $env;
        }

        $name = self::libraryName();
        $pkg = \dirname(__DIR__);   // .../sdks/php
        $repo = \dirname($pkg, 2);  // .../sdks/php -> .../sdks -> repo root

        $candidates = [$pkg . '/lib/' . $name];
        foreach (['debug', 'release'] as $profile) {
            $candidates[] = $repo . '/target/' . $profile . '/' . $name;
        }
        foreach ($candidates as $candidate) {
            if (\is_file($candidate)) {
                return $candidate;
            }
        }

        // Let the dynamic loader try. If it fails, an FFI\Exception naming the
        // soname is more useful than a guess at a path.
        return $name;
    }

    private static function libraryName(): string
    {
        switch (\PHP_OS_FAMILY) {
            case 'Windows':
                return 'patala_ffi.dll';
            case 'Darwin':
                return 'libpatala_ffi.dylib';
            default:
                return 'libpatala_ffi.so';
        }
    }
}
